### PR TITLE
NTP-533: clear filters button on search results page

### PR DESCRIPTION
**Pull request by stephenjmcneill**
## Context

Added clear filters button and end to end tests

## Changes proposed in this pull request

**Before:**
clear filters before

**After:**
Clear all filters after

## Guidance to review

Run through tuition partner journey, clear results and make sure results are updated. When no subjects are selected all filters should be closed

## Link to Jira ticket

https://dfedigital.atlassian.net/browse/NTP-533

## Things to check

- [x] Title is in the format "NTP-XXX: Short description" where NTP-XXX is the Jira ticket reference
- [x] Code and tests follow the [coding standards](/docs/coding-standards.md)
- [x] `dotnet format` has been run in the repository root
- [x] Test coverage of new code is at least 80%
- [x] All UI related acceptance criteria specified in the ticket have been added to the relevant [cypress test feature file(s)](/UI/cypress/e2e/)
- [ ] Unless explicitly mentioned in the ticket, all UI work should have been tested as working on desktop and mobile resolutions with JavaScript on and off
- [ ] Database migrations can be applied to the current codebase in main without causing exceptions
- [ ] Debug logging has been added after all logic decision points
- [ ] All [automated testing](/README.md#testing) including accessibility and security has been run against your code
- [ ] **PR deployment has been signed off by wider team**

### CODE REVIEW

File: UI/Pages/SearchResults.cshtml [hunk 1]
```diff
@@ -7,131 +7,131 @@
...
-	
-	</form>
+						Model.Data.AllSubjects[item].Select(x => ($"{item}-{x.Name}".ToSeoUrl(), $"{item}-{x.Name}", x.Name, x.Selected)).OrderBy(x=> x.Name)
+						)) " />
+                }
+
+                <govuk-radios asp-for="Data.TuitionType" class="govuk-radios--small">
+                    <govuk-radios-fieldset>
+                        <govuk-radios-fieldset-legend class="govuk-fieldset__legend--s">
+                            Type of tuitio
```
Comment (iterative-it) [inline UI/Pages/SearchResults.cshtml:43]: You've removed the 'n' from tuition here

File: UI/Pages/SearchResults.cshtml [hunk 2]
```diff
@@ -7,131 +7,131 @@
...
-				}
-			}
-		</div>
-	
-	</form>
+						Model.Data.AllSubjects[item].Select(x => ($"{item}-{x.Name}".ToSeoUrl(), $"{item}-{x.Name}", x.Name, x.Selected)).OrderBy(x=> x.Name)
+						)) " />
+                }
+
+                <govuk-radios asp-for="Data.TuitionType" class="govuk-radios--small">
```
Comment (iterative-it) [inline UI/Pages/SearchResults.cshtml:40]: Adding class `govuk-!-margin-bottom-3` will get this closer to the spacing of the design

File: UI/cypress/e2e/results.feature [hunk 3]
```diff
@@ -112,4 +112,22 @@ Feature: User is shown search results
...
+    And they will see the tuition type 'Any' is selected
+
+   Scenario: All subject filters are collapsed when clear all filters is clicked
+    Given a user has arrived on the 'Search results' page
+    When the ‘clear filters’ button as been selected
+    Then all subject filters are collapsed
+
+  Scenario: Results are updated when clear all filters is clicked
+    Given a user has arrived on the 'Search results' page
+    Then results are updated after filters are cleared
```
Comment (iterative-it) [inline UI/cypress/e2e/results.feature:133]: I think there is a missing scenario here which is also not in the ACs. Something like:

Scenario: Postcode is retained when clear all filters is clicked

File: UI/cypress/e2e/results.js [hunk 4]
```diff
@@ -109,4 +113,42 @@ Then("display all correct tuition partners that provide the selected subjects in
...
+    let countOfElements = 0;
+    cy.get('[data-testid="results-list-item"]').then($elements => {
+    countOfElements = $elements.length;
+
+    cy.get('[data-testid="clear-all-filters"]').click();
+
+    cy.get('[data-testid="results-list-item"]').should('have.length.greaterThan', countOfElements)
+  });
+});
+
```
Comment (iterative-it) [inline UI/cypress/e2e/results.js:143]: Please remove these extra line breaks

**Review by iterative-it — CHANGES_REQUESTED** [review]

File: UI/cypress/e2e/results.feature (hunk 3 above)
Comment (stephenjmcneill) [reply UI/cypress/e2e/results.feature]: Added new test

File: UI/cypress/e2e/results.js (hunk 4 above)
Comment (stephenjmcneill) [reply UI/cypress/e2e/results.js:140]: removed

File: UI/Pages/SearchResults.cshtml (hunk 2 above)
Comment (stephenjmcneill) [reply UI/Pages/SearchResults.cshtml:40]: Added class

**Comment by stephenjmcneill** [comment]
This ticket now has solution for ticket 570 included. Changing 'Any' tuition type to 'Both' in the search results page

File: UI/Pages/SearchResults.cshtml (hunk 1 above)
Comment (stephenjmcneill) [reply UI/Pages/SearchResults.cshtml]: Added back

File: UI/cypress/e2e/results.js [hunk 5]
```diff
@@ -109,4 +113,31 @@ Then("display all correct tuition partners that provide the selected subjects in
...
+    let countOfElements = 0;
+    cy.get('[data-testid="results-list-item"]').then($elements => {
+    countOfElements = $elements.length;
+    cy.get('[data-testid="clear-all-filters"]').click();
+    cy.get('[data-testid="results-list-item"]').should('have.length.greaterThan', countOfElements)
+    })
+});
+
+Then("the postcode search parameter remains", () => {
+    cy.get('[data-testid="postcode-input-box"]').should('not.be.null');
```
Comment (iterative-it) [inline UI/cypress/e2e/results.js:142]: This test doesn't fail if the code behind is altered to null out the postcode. The default for the input box is an empty string which is not null. Best to change this to look for the exact postcode entered to prove that it doesn't alter the postcode

File: UI/Pages/SearchResults.cshtml.cs [hunk 6]
```diff
@@ -35,6 +35,11 @@ public async Task OnGet(Query Data)
                 ModelState.AddModelError($"Data.{error.PropertyName}", error.ErrorMessage);
     }
 
+    public async Task OnGetClearAllFilters(string postcode)
+    {
+        this.Data = await mediator.Send(new Query { Postcode = postcode, Subjects = null, TuitionType = TuitionType.Both, KeyStages = null });
```
Comment (iterative-it) [inline UI/Pages/SearchResults.cshtml.cs:40]: There's no unit test covering this code. Please add a simple test or test covering resetting subjects, tuition type and key stages while maintaining postcode

**Review by iterative-it — CHANGES_REQUESTED** [review]

File: UI/cypress/e2e/results.js (hunk 5 above)
Comment (stephenjmcneill) [reply UI/cypress/e2e/results.js]: Changed

File: UI/Pages/SearchResults.cshtml.cs (hunk 6 above)
Comment (stephenjmcneill) [reply UI/Pages/SearchResults.cshtml.cs]: Unit test added

**Review by iterative-it — DISMISSED** [review]

**Review by iterative-it — APPROVED** [review]

**Pull request merged**